File: homeres/samples/viewmodel.ts
class HelloViewModel implements IHelloViewModelLocal {
    private name = 'world';
    private welcome = 'Hello, world!';

    readonly NameChanged = new RpcEvent<[]>();
    readonly WelcomeChanged = new RpcEvent<[]>();

    GetName(): string {
        return this.name;
    }

    async SetName(value: string): Promise<void> {
        if (this.name !== value) {
            this.name = value;
            this.welcome = `Hello, ${this.name}!`;
            await this.NameChanged.emit();
            await this.WelcomeChanged.emit();
        }
    }

    GetWelcome(): string {
        return this.welcome;
    }
}

File: homeres/samples/viewmodel.cpp
class HelloViewModel : public Object, public virtual IHelloViewModel
{
private:
    WString name = L"world";
    WString welcome = L"Hello, world!";

public:
    WString GetName() override
    {
        return name;
    }

    void SetName(const WString& value) override
    {
        if (name != value)
        {
            name = value;
            welcome = L"Hello, " + name + L"!";
            NameChanged();
            WelcomeChanged();
        }
    }

    WString GetWelcome() override
    {
        return welcome;
    }
};
